File: sample/src/commonMain/kotlin/io/github/linreal/cascade/screens/comments/CommentsScreen.kt
package io.github.linreal.cascade.screens.comments

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.imePadding
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.lazy.rememberLazyListState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import io.github.linreal.cascade.screens.TitledEditorTopBar
import io.github.linreal.cascade.theme.SampleEditorTheme
import io.github.linreal.cascade.ui.PageScaffold

/**
 * Comments sample — showcases [io.github.linreal.cascade.editor.ui.CascadeEditor]
 * as a chat-style comment composer. A feed of rich-text bubbles sits above a
 * composer whose formatting bar fades in on focus; Send appends the comment and
 * clears focus.
 */
@Composable
fun CommentsScreen(
    isDark: Boolean,
    onToggleTheme: () -> Unit,
    onBack: () -> Unit,
) {
    val editorTheme = if (isDark) SampleEditorTheme.dark() else SampleEditorTheme.light()
    val model = remember { CommentsScreenModel() }

    val listState = rememberLazyListState()
    LaunchedEffect(model.comments.size) {
        if (model.comments.isNotEmpty()) {
            listState.animateScrollToItem(model.comments.lastIndex)
        }
    }

    PageScaffold {
        TitledEditorTopBar(
            title = "Comments",
            isDark = isDark,
            onBack = onBack,
            onToggleTheme = onToggleTheme,
        )

        Box(
            modifier = Modifier
                .fillMaxWidth()
                .weight(1f)
                .imePadding(),
        ) {
            LazyColumn(
                state = listState,
                modifier = Modifier.fillMaxSize(),
                contentPadding = PaddingValues(
                    top = 16.dp,
                    bottom = FeedBottomInset,
                ),
                verticalArrangement = Arrangement.spacedBy(18.dp),
            ) {
                items(items = model.comments, key = { it.id }) { comment ->
                    CommentBubble(comment)
                }
            }

            CommentComposer(
                model = model,
                editorTheme = editorTheme,
                modifier = Modifier
                    .align(Alignment.BottomCenter)
                    .fillMaxWidth(),
            )
        }
    }
}

private val FeedBottomInset = 150.dp

@Composable
private fun CommentBubble(comment: Comment) {
    val bubbleColor: Color
    val textColor: Color
    val inlineCodeBackground: Color
    val linkColor: Color
    if (comment.isOwn) {
        bubbleColor = MaterialTheme.colorScheme.primary
        textColor = MaterialTheme.colorScheme.onPrimary
        inlineCodeBackground = MaterialTheme.colorScheme.onPrimary.copy(alpha = 0.18f)
        linkColor = MaterialTheme.colorScheme.onPrimary
    } else {
        bubbleColor = MaterialTheme.colorScheme.surface
        textColor = MaterialTheme.colorScheme.onSurface
        inlineCodeBackground = MaterialTheme.colorScheme.onSurface.copy(alpha = 0.08f)
        linkColor = MaterialTheme.colorScheme.primary
    }

    val annotated = rememberCommentAnnotatedString(
        text = comment.text,
        spans = comment.spans,
        inlineCodeBackground = inlineCodeBackground,
        linkColor = linkColor,
    )

    val bubbleShape = if (comment.isOwn) {
        RoundedCornerShape(topStart = 16.dp, topEnd = 4.dp, bottomEnd = 16.dp, bottomStart = 16.dp)
    } else {
        RoundedCornerShape(topStart = 4.dp, topEnd = 16.dp, bottomEnd = 16.dp, bottomStart = 16.dp)
    }

    Row(
        modifier = Modifier.fillMaxWidth(),
        horizontalArrangement = if (comment.isOwn) Arrangement.End else Arrangement.Start,
    ) {
        if (comment.isOwn) {
            Spacer(modifier = Modifier.widthIn(min = 40.dp))
        } else {
            Avatar(comment)
            Spacer(modifier = Modifier.size(11.dp))
        }

        Column(
            horizontalAlignment = if (comment.isOwn) Alignment.End else Alignment.Start,
            modifier = Modifier.widthIn(max = 300.dp),
        ) {
            Row(verticalAlignment = Alignment.CenterVertically) {
                Text(
                    text = comment.authorName,
                    fontSize = 14.sp,
                    fontWeight = FontWeight.SemiBold,
                    color = MaterialTheme.colorScheme.onBackground,
                )
                Spacer(modifier = Modifier.size(8.dp))
                Text(
                    text = comment.timestamp,
                    fontSize = 12.sp,
                    color = MaterialTheme.colorScheme.onBackground.copy(alpha = 0.5f),
                )
            }
            Spacer(modifier = Modifier.height(6.dp))
            Text(
                text = annotated,
                fontSize = 15.sp,
                lineHeight = 22.sp,
                color = textColor,
                modifier = Modifier
                    .clip(bubbleShape)
                    .background(bubbleColor)
                    .padding(horizontal = 14.dp, vertical = 11.dp),
            )
        }

        if (comment.isOwn) {
            Spacer(modifier = Modifier.size(11.dp))
            Avatar(comment)
        } else {
            Spacer(modifier = Modifier.widthIn(min = 40.dp))
        }
    }
}

@Composable
private fun Avatar(comment: Comment) {
    Box(
        modifier = Modifier
            .size(36.dp)
            .clip(CircleShape)
            .background(comment.avatarColor),
        contentAlignment = Alignment.Center,
    ) {
        Text(
            text = comment.initials,
            fontSize = 13.sp,
            fontWeight = FontWeight.SemiBold,
            color = Color.White,
        )
    }
}

File: iosNativeSample/iosNativeSample/Screens/CommentsScreen.swift
import SwiftUI
import CascadeEditor

/// Editor-as-composer demo: a feed of rich-text bubbles above a pinned
/// composer whose formatting bar fades in while the editor is focused.
struct CommentsScreen: View {
    @Binding var isDark: Bool
    @Environment(\.dismiss) private var dismiss
    @StateObject private var editor: EditorScreenModel
    @StateObject private var model: CommentsModel

    init(isDark: Binding<Bool>) {
        _isDark = isDark
        let editorModel = EditorScreenModel(
            configuration: .standard(
                isDark: isDark.wrappedValue,
                toolbarMode: CascadeToolbarMode.none,
                slashCommandsEnabled: false,
                blockSelectionEnabled: false,
                blockDraggingEnabled: false
            )
        )
        _editor = StateObject(wrappedValue: editorModel)
        _model = StateObject(wrappedValue: CommentsModel(editor: editorModel))
    }

    private var theme: AppTheme { AppTheme.theme(isDark: isDark) }

    var body: some View {
        VStack(spacing: 0) {
            TitledEditorTopBar(
                theme: theme,
                title: "Comments",
                onBack: { dismiss() },
                onToggleTheme: { isDark.toggle() }
            )
            feed
            composer
        }
        .padding(.horizontal, 12)
        .background(theme.background)
        .onChange(of: isDark) { _, newValue in
            editor.setDarkMode(newValue)
        }
    }

    private var feed: some View {
        ScrollViewReader { proxy in
            ScrollView {
                LazyVStack(spacing: 18) {
                    ForEach(model.comments) { comment in
                        CommentBubble(theme: theme, comment: comment)
                            .id(comment.id)
                    }
                }
                .padding(.top, 16)
                .padding(.bottom, 20)
            }
            .onChange(of: model.comments.count) { _, _ in
                if let last = model.comments.last {
                    withAnimation {
                        proxy.scrollTo(last.id, anchor: .bottom)
                    }
                }
            }
        }
    }

    private var composer: some View {
        VStack(spacing: 0) {
            if editor.toolbarState.focused {
                FormattingBar(theme: theme, editor: editor)
                    .transition(.opacity.combined(with: .move(edge: .bottom)))
            }
            HStack(spacing: 8) {
                CascadeEditorHost(model: editor)
                    .frame(height: 64)
                    .clipShape(RoundedRectangle(cornerRadius: 20))
                SendButton(theme: theme, enabled: model.canSend) {
                    model.send()
                }
            }
            .padding(EdgeInsets(top: 4, leading: 6, bottom: 4, trailing: 8))
            .background(theme.surface)
            .clipShape(RoundedRectangle(cornerRadius: 24))
        }
        .padding(.top, 8)
        .padding(.bottom, 12)
        .animation(.default, value: editor.toolbarState.focused)
    }
}

private struct CommentBubble: View {
    let theme: AppTheme
    let comment: Comment

    var body: some View {
        let bubbleColor = comment.isOwn ? theme.primary : theme.surface
        let textColor = comment.isOwn ? theme.onPrimary : theme.onSurface
        let inlineCodeBackground = UIColor(textColor).withAlphaComponent(comment.isOwn ? 0.18 : 0.08)
        let linkColor = comment.isOwn ? UIColor(theme.onPrimary) : UIColor(theme.primary)

        HStack(alignment: .top, spacing: 11) {
            if comment.isOwn {
                Spacer(minLength: 40)
            } else {
                Avatar(comment: comment)
            }

            VStack(alignment: comment.isOwn ? .trailing : .leading, spacing: 6) {
                HStack(spacing: 8) {
                    Text(comment.authorName)
                        .font(.system(size: 14, weight: .semibold))
                        .foregroundStyle(theme.onBackground)
                    Text(comment.timestamp)
                        .font(.system(size: 12))
                        .foregroundStyle(theme.onBackground.opacity(0.5))
                }
                Text(
                    commentAttributedString(
                        text: comment.text,
                        spans: comment.spans,
                        textColor: UIColor(textColor),
                        inlineCodeBackground: inlineCodeBackground,
                        linkColor: linkColor
                    )
                )
                .font(.system(size: 15))
                .lineSpacing(4)
                .padding(.horizontal, 14)
                .padding(.vertical, 11)
                .background(bubbleColor)
                .clipShape(BubbleShape(isOwn: comment.isOwn))
            }
            .frame(maxWidth: 300, alignment: comment.isOwn ? .trailing : .leading)

            if comment.isOwn {
                Avatar(comment: comment)
            } else {
                Spacer(minLength: 40)
            }
        }
        .frame(maxWidth: .infinity, alignment: comment.isOwn ? .trailing : .leading)
    }
}

private struct BubbleShape: Shape {
    let isOwn: Bool

    func path(in rect: CGRect) -> Path {
        let sharpCorner: UIRectCorner = isOwn ? .topRight : .topLeft
        let rounded = UIBezierPath(
            roundedRect: rect,
            byRoundingCorners: UIRectCorner.allCorners.subtracting(sharpCorner),
            cornerRadii: CGSize(width: 16, height: 16)
        )
        let sharp = UIBezierPath(
            roundedRect: rect,
            byRoundingCorners: sharpCorner,
            cornerRadii: CGSize(width: 4, height: 4)
        )
        var path = Path(rounded.cgPath)
        path = path.intersection(Path(sharp.cgPath))
        return path
    }
}

private struct Avatar: View {
    let comment: Comment

    var body: some View {
        Text(comment.initials)
            .font(.system(size: 13, weight: .semibold))
            .foregroundStyle(.white)
            .frame(width: 36, height: 36)
            .background(comment.avatarColor)
            .clipShape(Circle())
    }
}

/// Formatting bar shown while the composer is focused. Buttons drive the
/// controller's toolbar actions; SwiftUI buttons never become first responder,
/// so tapping them keeps the editor focused.
private struct FormattingBar: View {
    let theme: AppTheme
    @ObservedObject var editor: EditorScreenModel

    var body: some View {
        HStack(spacing: 6) {
            formatButton("bold", "Bold", editor.toolbarState.bold) { editor.controller.toggleBold() }
            formatButton("italic", "Italic", editor.toolbarState.italic) { editor.controller.toggleItalic() }
            formatButton("underline", "Underline", editor.toolbarState.underline) { editor.controller.toggleUnderline() }
            Rectangle()
                .fill(theme.onSurface.opacity(0.12))
                .frame(width: 1, height: 20)
                .padding(.horizontal, 3)
            formatButton("strikethrough", "Strikethrough", editor.toolbarState.strikeThrough) {
                editor.controller.toggleStrikeThrough()
            }
            formatButton("chevron.left.forwardslash.chevron.right", "Inline code", editor.toolbarState.inlineCode) {
                editor.controller.toggleInlineCode()
            }
            Spacer()
        }
        .padding(.horizontal, 4)
        .padding(.vertical, 8)
    }

    private func formatButton(
        _ systemImage: String,
        _ label: String,
        _ state: CascadeStyleState,
        action: @escaping () -> Void
    ) -> some View {
        let enabled = editor.toolbarState.canFormat
        let active = state == CascadeStyleState.active || state == CascadeStyleState.mixed
        let background: Color = !enabled ? .clear : (active ? theme.primary.opacity(0.16) : theme.surface)
        let tint: Color = !enabled
            ? theme.onSurface.opacity(0.3)
            : (active ? theme.primary : theme.onSurface.opacity(0.7))

        return Button(action: action) {
            Image(systemName: systemImage)
                .font(.system(size: 14, weight: .medium))
                .foregroundStyle(tint)
                .frame(width: 38, height: 38)
                .background(background)
                .clipShape(RoundedRectangle(cornerRadius: 9))
        }
        .buttonStyle(.plain)
        .disabled(!enabled)
        .accessibilityLabel(label)
    }
}

private struct SendButton: View {
    let theme: AppTheme
    let enabled: Bool
    let action: () -> Void

    var body: some View {
        Button(action: action) {
            Image(systemName: "paperplane.fill")
                .font(.system(size: 17, weight: .medium))
                .foregroundStyle(theme.onPrimary)
                .frame(width: 44, height: 44)
                .background(theme.primary.opacity(enabled ? 1 : 0.3))
                .clipShape(Circle())
        }
        .buttonStyle(.plain)
        .disabled(!enabled)
        .accessibilityLabel("Send comment")
    }
}
